src/xtgeo/grid3d/_grdecl_grid.py

```diff
@@ -0,0 +1,644 @@
+    def valid_mapaxes(mapaxes):
+        y_line = mapaxes.y_line
+        x_line = mapaxes.x_line
+        origin = mapaxes.origin
+        x_axis = np.array(x_line) - origin
+        y_axis = np.array(y_line) - origin
+
+        return np.linalg.norm(x_axis) > 1e-5 and np.linalg.norm(y_axis) > 1e-5
+
+    def to_file(self, filename):
+        with open(filename, "w") as filestream:
+            keywords = [
+                ("SPECGRID", self.specgrid.to_grdecl()),
+                ("MAPAXES", self.mapaxes.to_grdecl() if self.mapaxes else None),
+                ("MAPUNITS", [self.mapunits] if self.mapunits else None),
+                ("GRIDUNIT", self.gridunit.to_grdecl() if self.gridunit else None),
+                ("GDORIENT", self.gdorient.to_grdecl() if self.gdorient else None),
+                ("COORD", self.coord),
+                ("ZCORN", self.zcorn),
+                ("ACTNUM", self.actnum),
+            ]
+            for kw, values in keywords:
+                if values is None:
+                    continue
+                filestream.write(kw)
+                if values is not None:
+                    filestream.write("\n")
+                    for value in values:
+                        filestream.write(" ")
+                        filestream.write(str(value))
+                filestream.write("\n /\n")
+
+    @property
+    def dimensions(self):
+        return (self.specgrid.ndivix, self.specgrid.ndiviy, self.specgrid.ndiviz)
+
+    def _check_xtgeo_compatible(self):
+        if self.specgrid.coordinate_type == CoordinateType.CYLINDRICAL:
+            raise NotImplementedError(
+                "Xtgeo does not currently support cylindrical coordinate systems"
+            )
+        if self.specgrid.numres != 1:
+            raise NotImplementedError(
+                "Xtgeo does not currently support multiple reservoirs"
+            )
+        if self.gridunit and self.gridunit.grid_relative == GridRelative.MAP:
+            raise NotImplementedError(
+                "Xtgeo does not currently support conversion of map relative grid units"
+            )
+
+    def xtgeo_coord(self):
+        """
+        Returns:
+            coord in xtgeo format.
+        """
+        self._check_xtgeo_compatible()
+        nx, ny, _ = self.dimensions
+
+        xtgeo_coord = np.swapaxes(self.coord.reshape((ny + 1, nx + 1, 6)), 0, 1).astype(
+            np.float64
+        )
+        if self.mapaxes:
+            self.transform_xtgeo_coord_by_mapaxes(xtgeo_coord)
+        return np.ascontiguousarray(xtgeo_coord)
+
+    def xtgeo_actnum(self):
+        """
+        Returns:
+            actnum in xtgeo format.
+        """
+        self._check_xtgeo_compatible()
+        nx, ny, nz = self.dimensions
+        if self.actnum is None:
+            return np.ones(shape=(nx, ny, nz), dtype=np.int32)
+        actnum = self.actnum.reshape((nx, ny, nz), order="F")
+        return np.ascontiguousarray(actnum)
+
+    def xtgeo_zcorn(self):
```

Comment:
I would like to see some simple tests of this. It looks to be well integration tested, but would be nice with the simplest possible grid spelled out, something like a 2x2 grid for example.

Reply:
Even at a 2x2x2 grid it is pretty difficult to read what is going on but I'll add one in any case :)

Reply:
Added a simpler test, but I think it gets quite hairy even with the simplest values I could think of (that are still distinct).

Reply:
Yes, it is quite hairy, but it as least easier to read what is going on I think, thanks for adding the test!